packages/demos/stories/DragDrop.tsx

```diff
@@ -0,0 +1,200 @@
+/**
+ * © 2020 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayList from '@clayui/list';
+import {XYCoord} from 'dnd-core';
+import React, {useState, useCallback, useRef} from 'react';
+import {useDrag, useDrop, DropTargetMonitor} from 'react-dnd';
+
+interface IDraggableListItemProps {
+	/**
+	 * List Item unique identifier
+	 */
+	id: any;
+
+	/**
+	 * The text displayed in the List Item
+	 */
+	text: string;
+
+	/**
+	 * Index of the List Item in order to keep track of it's positioning within the array
+	 */
+	index: number;
+
+	/**
+	 * Function that handles the moving of List Items
+	 */
+	onMove: (dragIndex: number, hoverIndex: number) => void;
+}
+
+interface IDragItem {
+	/**
+	 * List Item unique identifier
+	 */
+	id: string;
+
+	/**
+	 * Index of the List Item in order to keep track of it's positioning within the array
+	 */
+	index: number;
+
+	/**
+	 * Type of data that is allowed to be dragged
+	 */
+	type: string;
+}
+
+const DraggableListItem: React.FunctionComponent<IDraggableListItemProps> = ({
+	id,
+	index,
+	onMove,
+	text,
+}) => {
+	const ref = useRef<HTMLDivElement>(null);
+
+	const [, drop] = useDrop({
+		accept: 'listItem',
+		hover(item: IDragItem, monitor: DropTargetMonitor) {
+			const dragIndex = item.index;
+
+			const hoverIndex = index;
+
+			if (!ref.current || dragIndex === hoverIndex) {
+				return;
+			}
+
+			const hoverBoundingRect = ref.current!.getBoundingClientRect();
+
+			const verticalMiddle =
+				(hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
+
+			const mousePosition = monitor.getClientOffset();
+
+			const pixelsToTop =
+				(mousePosition as XYCoord).y - hoverBoundingRect.top;
+
+			const draggingDownwards =
+				dragIndex < hoverIndex && pixelsToTop < verticalMiddle / 2;
+
+			if (draggingDownwards) {
+				return;
+			}
+
+			const draggingUpwards =
+				dragIndex > hoverIndex && pixelsToTop > verticalMiddle * 1.5;
+
+			if (draggingUpwards) {
+				return;
+			}
+
+			onMove(dragIndex, hoverIndex);
+
+			item.index = hoverIndex;
+		},
+	});
+
+	const [{isDragging, itemBeingDragged}, drag] = useDrag({
+		collect: (monitor: any) => ({
+			isDragging: monitor.isDragging(),
+			itemBeingDragged: monitor.getItem() || {id: 0},
+		}),
+		item: {id, index, type: 'listItem'},
+	});
+
+	const isItemBeingDragged = itemBeingDragged.id === id;
+
+	const draggingStyle = {
+		backgroundColor: '#FFF8C9',
+		border: '2px solid #4B9FFF',
+	};
+
+	drag(drop(ref));
+
+	const opacity = isDragging ? 0 : 1;
+
+	return (
+		<div
```

Comment:
Lets use `<ClayList>` instead of a div. This way we can make it look like a proper clay list.

Reply:
Or actually, `<ClayList>` should probably be the parent container to `{listItems.map` in `DraggableList` and then this component primary return should be `<ClayList.Item>`. This was we properly match the semantic markup from clay

Reply:
```
<ClayList>
  <div ref={ref}>
    <ClayList.Item>
```
Is as semantic I can get it without adding `forwardRef` to the ClayList.Item

Reply:
Oh gotcha, this is a good chance for us to add a `fix: ` of adding forwardRef to `ClayList.Item`. I'm sure others will also have a reason to use forwardRefs with claylist.